sql/memo/exec_builder.go

```diff
@@ -133,6 +133,94 @@ func (b *ExecBuilder) buildLookupJoin(j *LookupJoin, input sql.Schema, children
 	return plan.NewJoin(left, right, j.Op, filters).WithScopeLen(j.g.m.scopeLen), nil
 }
 
+func (b *ExecBuilder) buildSlidingRange(sr *SlidingRange, leftSch, rightSch sql.Schema, children ...sql.Node) (ret sql.Node, err error) {
+	switch n := children[0].(type) {
+	case *plan.TableAlias:
+		ret, err = b.buildSlidingRange(sr, leftSch, rightSch, n.Child)
+		ret = plan.NewTableAlias(n.Name(), ret)
+	case *plan.Distinct:
+		ret, err = b.buildSlidingRange(sr, leftSch, rightSch, n.Child)
+		ret = plan.NewDistinct(ret)
+	case *plan.Filter:
+		ret, err = b.buildSlidingRange(sr, leftSch, rightSch, n.Child)
+		ret = plan.NewFilter(n.Expression, ret)
+	case *plan.Project:
+		ret, err = b.buildSlidingRange(sr, leftSch, rightSch, n.Child)
+		ret = plan.NewProject(n.Projections, ret)
+	case *plan.Limit:
+		ret, err = b.buildSlidingRange(sr, leftSch, rightSch, n.Child)
+		ret = plan.NewLimit(n.Limit, ret)
+	default:
+		var childNode sql.Node
+		if sr.RightIndex != nil {
+			childNode, err = b.buildIndexScan(sr.RightIndex, rightSch, n)
+		} else {
+			sortExpr, err := b.buildScalar(*sr.MinExpr, rightSch)
+			if err != nil {
+				return nil, err
+			}
+			sf := []sql.SortField{{
+				Column:       sortExpr,
+				Order:        sql.Ascending,
+				NullOrdering: sql.NullsFirst,
+			}}
+			childNode = plan.NewSort(sf, n)
+		}
+
+		if err != nil {
+			return nil, err
+		}
+		ret, err = plan.NewSlidingRange(
+			childNode,
+			leftSch,
+			rightSch,
+			sr.ValueCol.Gf.Name(),
+			sr.MinColRef.Gf.Name(),
+			sr.MaxColRef.Gf.Name(),
+			sr.RangeClosedOnLowerBound,
+			sr.RangeClosedOnUpperBound)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+func (b *ExecBuilder) buildSlidingRangeJoin(j *SlidingRangeJoin, input sql.Schema, children ...sql.Node) (sql.Node, error) {
+	leftSch := input[:len(input)-len(j.Right.RelProps.OutputCols())]
+	rightSch := input[len(j.Left.RelProps.OutputCols()):]
+
+	var left sql.Node
+	var err error
+	if j.SlidingRange.LeftIndex != nil {
+		left, err = b.buildIndexScan(j.SlidingRange.LeftIndex, input, children[0])
+		if err != nil {
+			return nil, err
+		}
+	} else {
```

Comment:
splitting the sort out into a separate node probably makes sense at some point, otherwise debugging/costing might be a bit tricker

Reply:
Right now if we need to the coster can easily tell whether the plan will use an IndexScan or a Sort based on whether the Index field is set, and the described plan is clearly different, showing either the IndexScan or the Sort.

But this is definitely something to keep in mind when refactoring this.